ipfs-retriever/lib/store.js

```diff
@@ -0,0 +1,392 @@
+  )
+
+  return { serviceProviderId, serviceUrl, dataSetId, pieceId, ipfsRootCid }
+}
+
+/**
+ * Retrieves the provider and data set id for a given root CID.
+ *
+ * @param {Pick<Env, 'DB'>} env - Cloudflare Worker environment with D1 DB
+ *   binding
+ * @param {string} payerAddress - The address of the client paying for the
+ *   request
+ * @param {string} ipfsRootCid - The IPFS Root CID to look up
+ * @returns {Promise<{
+ *   serviceProviderId: string
+ *   serviceUrl: string
+ *   dataSetId: string
+ *   pieceId: string
+ * }>}
+ */
+export async function getStorageProviderAndValidatePayerByWalletAndCid(
+  env,
+  payerAddress,
+  ipfsRootCid,
+) {
+  const query = `
+   SELECT
+     pieces.id as piece_id,
+     pieces.data_set_id,
+     pieces.ipfs_root_cid,
+     data_sets.service_provider_id,
+     data_sets.payer_address,
+     data_sets.with_cdn,
+     data_sets.with_ipfs_indexing,
+     service_providers.service_url,
+     wallet_details.is_sanctioned
+   FROM pieces
+   LEFT OUTER JOIN data_sets
+     ON pieces.data_set_id = data_sets.id
+   LEFT OUTER JOIN service_providers
+     ON data_sets.service_provider_id = service_providers.id
+   LEFT OUTER JOIN wallet_details
+     ON data_sets.payer_address = wallet_details.address
+   WHERE pieces.ipfs_root_cid = ?
+ `
+
+  const results = /**
+   * @type {{
+   *   piece_id: string
+   *   data_set_id: string
+   *   ipfs_root_cid: string
+   *   service_provider_id: string
+   *   payer_address: string | undefined
+   *   with_cdn: number | undefined
+   *   with_ipfs_indexing: number | undefined
+   *   service_url: string | undefined
+   *   is_sanctioned: number | undefined
+   * }[]}
+   */ (
+    /** @type {any[]} */ (
+      (await env.DB.prepare(query).bind(ipfsRootCid).all()).results
+    )
+  )
+
+  return validateQueryResultsAndGetProvider({
+    results,
+    payerAddress,
+    lookupKey: `IPFS Root CID '${ipfsRootCid}'`,
+  })
+}
+
+/**
+ * Retrieves the provider info for a given data set ID and piece ID.
+ *
+ * @param {Pick<Env, 'DB'>} env - Cloudflare Worker environment with D1 DB
+ *   binding
+ * @param {string} dataSetId - The data set ID
+ * @param {string} pieceId - The piece ID
+ * @returns {Promise<{
+ *   serviceProviderId: string
+ *   serviceUrl: string
+ *   dataSetId: string
+ *   pieceId: string
+ *   ipfsRootCid: string
+ * }>}
+ */
+export async function getStorageProviderAndValidatePayerByDataSetAndPiece(
+  env,
+  dataSetId,
+  pieceId,
+) {
+  if (dataSetId === '9999' && pieceId === '9999') {
+    // Special case for testing purposes only
+    return {
+      serviceProviderId: '9999',
+      serviceUrl: 'https://frisbii.fly.dev/',
+      dataSetId,
+      pieceId,
+      ipfsRootCid:
+        'bafybeiagrjpf2rwth5oylc64czsrz2jm7a4fgo67b2luygqjrivjbswuku',
+    }
+  }
```

Comment:
@juliangruber I believe you wanted to remove the hard-coded combination pointing to our Frisbii. I think we should remove this block as well. Thoughts?